internal/status/httproute.go

```diff
@@ -74,3 +59,42 @@ func prepareHTTPRouteStatus(
 		},
 	}
 }
+
+func convertRouteConditions(
+	routeConds []conditions.RouteCondition,
+	observedGeneration int64,
+	transitionTime metav1.Time,
+) []metav1.Condition {
+	conds := make([]metav1.Condition, 0, len(routeConds))
```

Comment:
here are the results on my machine (convertRouteConditions2 uses a traditional loop)
```
BenchmarkConvert/convertRouteConditions_with_1_conditions-4             12886671                83.75 ns/op           96 B/op          1 allocs/op
BenchmarkConvert/convertRouteConditions2_with_1_conditions-4            12123184                83.38 ns/op           96 B/op          1 allocs/op
BenchmarkConvert/convertRouteConditions_with_2_conditions-4              7348525               153.2 ns/op           192 B/op          1 allocs/op
BenchmarkConvert/convertRouteConditions2_with_2_conditions-4             8000790               148.1 ns/op           192 B/op          1 allocs/op
BenchmarkConvert/convertRouteConditions_with_5_conditions-4              3855415               320.3 ns/op           480 B/op          1 allocs/op
BenchmarkConvert/convertRouteConditions2_with_5_conditions-4             3862740               304.5 ns/op           480 B/op          1 allocs/op
BenchmarkConvert/convertRouteConditions_with_10_conditions-4             1970270               616.3 ns/op          1024 B/op          1 allocs/op
BenchmarkConvert/convertRouteConditions2_with_10_conditions-4            2042850               606.9 ns/op          1024 B/op          1 allocs/op
BenchmarkConvert/convertRouteConditions_with_25_conditions-4              672940              1585 ns/op            2688 B/op          1 allocs/op
BenchmarkConvert/convertRouteConditions2_with_25_conditions-4             662079              1540 ns/op            2688 B/op          1 allocs/op
BenchmarkConvert/convertRouteConditions_with_50_conditions-4              396487              2788 ns/op            4864 B/op          1 allocs/op
BenchmarkConvert/convertRouteConditions2_with_50_conditions-4             452823              2658 ns/op            4864 B/op          1 allocs/op
BenchmarkConvert/convertRouteConditions_with_100_conditions-4             190802              5393 ns/op            9728 B/op          1 allocs/op
BenchmarkConvert/convertRouteConditions2_with_100_conditions-4            225346              5111 ns/op            9728 B/op          1 allocs/op
PASS
ok      github.com/nginxinc/nginx-kubernetes-gateway/internal/status    23.075s
```

I'd prefer the append style instead of indices when there is no much difference in performance.

Source:

```go
func convertRouteConditions2(
	routeConds []conditions.RouteCondition,
	observedGeneration int64,
	transitionTime metav1.Time,
) []metav1.Condition {
	conds := make([]metav1.Condition, len(routeConds), len(routeConds))

	for i := 0; i < len(routeConds); i++ {
		conds[i] = metav1.Condition{
			Type:               string(routeConds[i].Type),
			Status:             routeConds[i].Status,
			ObservedGeneration: observedGeneration,
			LastTransitionTime: transitionTime,
			Reason:             string(routeConds[i].Reason),
			Message:            routeConds[i].Message,
		}
	}

	return conds
}
```

```go
func generateConditions(n int) []conditions.RouteCondition {
	conds := make([]conditions.RouteCondition, 0, n)

	for i := 0; i < n; i++ {
		conds = append(conds, conditions.RouteCondition{
			Type:    v1beta1.RouteConditionType(fmt.Sprintf("Test%d", i)),
			Status:  metav1.ConditionTrue,
			Reason:  v1beta1.RouteConditionReason(fmt.Sprintf("reason%d", i)),
			Message: fmt.Sprintf("some about the condtion message%d", i),
		})
	}
	return conds
}

func BenchmarkConvert(b *testing.B) {
	counts := []int{
		1,
		2,
		5,
		10,
		25,
		50,
		100,
	}

	for _, count := range counts {
		conds := generateConditions(count)
		time := metav1.Now()

		b.Run(fmt.Sprintf("convertRouteConditions with %d conditions", count), func(b *testing.B) {
			bench(b, conds, convertRouteConditions, count, time)
		})

		b.Run(fmt.Sprintf("convertRouteConditions2 with %d conditions", count), func(b *testing.B) {
			bench(b, conds, convertRouteConditions2, count, time)
		})
	}
}

type convertFunc func(routeConds []conditions.RouteCondition, observedGeneration int64, transitionTime metav1.Time) []metav1.Condition

func bench(b *testing.B, conds []conditions.RouteCondition, convert convertFunc, n int, t metav1.Time) {
	for i := 0; i < b.N; i++ {
		res := convert(conds, 1, t)
		if len(res) != n {
			b.Fatalf("expected %d conditinos, got %d", n, len(res))
		}
	}
}
```

Reply:
discussed offline. our new guideline:
* if slice length and capacity is known at initialization, use index assignment over the cost of append. Unless taking values from a map (to avoid awkward indexing)
* append should be used for dynamically sized containers to avoid nil exceptions